crates/ty_python_semantic/resources/mdtest/import/workspaces.md

```diff
@@ -0,0 +1,660 @@
+
+`my-tests/main.py`:
+
+```py
+# TODO: there should be no errors in this file
+
+# error: [unresolved-import]
+from .mod1 import x
+
+# error: [unresolved-import]
+from . import mod2
+import mod3
+
+reveal_type(x)  # revealed: Unknown
+reveal_type(mod2.y)  # revealed: Unknown
+reveal_type(mod3.z)  # revealed: int
+```
+
+`my-tests/mod1.py`:
+
+```py
+x: int = 1
+```
+
+`my-tests/mod2.py`:
+
+```py
+y: int = 2
+```
+
+`mod3.py`:
+
+```py
+z: int = 2
+```
+
+### Current Directory Is Invalid Package Name
+
+Relative and absolute imports should resolve fine in a dir that isn't a valid package name, even if
+it contains an `__init__.py`:
+
+`my-tests/__init__.py`:
+
+```py
+```
+
+`my-tests/main.py`:
+
+```py
+# TODO: there should be no errors in this file
+
+# error: [unresolved-import]
+from .mod1 import x
+
+# error: [unresolved-import]
+from . import mod2
+import mod3
+
+reveal_type(x)  # revealed: Unknown
+reveal_type(mod2.y)  # revealed: Unknown
+reveal_type(mod3.z)  # revealed: int
+```
+
+`my-tests/mod1.py`:
+
+```py
+x: int = 1
+```
+
+`my-tests/mod2.py`:
+
+```py
+y: int = 2
+```
+
+`mod3.py`:
+
+```py
+z: int = 2
+```
+
+## Multiple Projects
+
+It's common for a monorepo to define many separate projects that may or may not depend on eachother
+and are stitched together with a package manager like `uv` or `poetry`, often as editables. In this
+case, especially when running as an LSP, we want to be able to analyze all of the projects at once,
+allowing us to reuse results between projects, without getting confused about things that only make
+sense when analyzing the project separately.
+
+The following tests will feature two projects, `a` and `b` where the "real" packages are found under
+`src/` subdirectories (and we've been configured to understand that), but each project also contains
+other python files in their roots or subdirectories that contains python files which relatively
+import eachother and also absolutely import the main package of the project. All of these imports
+*should* resolve.
+
+Often the fact that there is both an `a` and `b` project seemingly won't matter, but many possible
+solutions will misbehave under these conditions, as e.g. if both define a `main.py` and test code
+has `import main`, we need to resolve each project's main as appropriate.
+
+One key hint we will have in these situations is the existence of a `pyproject.toml`, so the
+following examples include them in case they help.
+
+### Tests Directory With Overlapping Names
+
+Here we have fairly typical situation where there are two projects `aproj` and `bproj` where the
+"real" packages are found under `src/` subdirectories, but each project also contains a `tests/`
+directory that contains python files which relatively import eachother and also absolutely import
+the package they test. All of these imports *should* resolve.
+
+```toml
+[environment]
+# This is similar to what we would compute for installed editables
```

Comment:
Those tests work because we also apply our default `environment.root` paths in mdtests?

Reply:
I believe so